shapely/coordinates.py

```diff
@@ -54,20 +69,252 @@ def transform(geometry, transformation, include_z=False):
+    rebuild: bool, default False
+        If False, output geometries will preserve the coordinate number and
+        dimnensionality as the input geometries (but in include_z=False,
+        two-dimension geometries are always returned).
+        Otherwise, output geometries will get rebuilt from the transformation
+        output, so they might have different number of coordinate or
+        a different coordiante-dimensionality.
+
+    Examples
+    --------
+    >>> from shapely import LineString, Point
+    >>> transform_extra(Point(0, 0), lambda x: x + 1)
+    <POINT (1 1)>
+    >>> transform_extra(LineString([(2, 2), (4, 4)]), lambda x: x * [2, 3])
+    <LINESTRING (4 6, 8 12)>
+    >>> transform_extra(None, lambda x: x) is None
+    True
+    >>> transform_extra([Point(0, 0), None], lambda x: x).tolist()
+    [<POINT (0 0)>, None]
+
+    By default, the third dimension is ignored:
+
+    >>> transform_extra(Point(0, 0, 0), lambda x: x + 1)
+    <POINT (1 1)>
+    >>> transform_extra(Point(0, 0, 0), lambda x: x + 1, include_z=True)
+    <POINT Z (1 1 1)>
+
+    An identity function applicable to both types of input with interleaved=False
+
+    >>> g1 = LineString([(1, 2), (3, 4)])
+
+    >>> transform_extra(g1, lambda x, y, z=None: tuple(filter(None, [x, y, z])), interleaved=False)
+    <LINESTRING (1 2, 3 4)>
+
+    example of another lambda expression:
+
+    >>> transform_extra(g1, lambda x, y, z=None: (x+1.0, y+1.0), interleaved=False)
+    <LINESTRING (2 3, 4 5)>
+
+    Using pyproj >= 2.1, the following example will accurately project Shapely geometries
+    It transforms from EPSG:4326 (WGS84 log/lat) to EPSG:32618 (WGS84 UTM 18 North)
+    Note: always_xy kwarg is required as Shapely geometries only support X,Y coordinate ordering.
+
+    >>> try:
+    ...     import pyproj
+    ...     project = pyproj.Transformer.from_crs(4326, 32618, always_xy=True).transform
+    ...     p = transform_extra(Point(-75, 50), project, interleaved=False)
+    ...     assert (round(p.x), round(p.y)) == (500000, 5538631)
+    ... except ImportError:
+    ...     pass
+    """
+    if include_z is None:
+        include_z = shapely.get_coordinate_dimension(geometry) == 3
+    if rebuild:
+        transform_wrapper = (
+            transform_rebuild if interleaved else transform_rebuild_planar
+        )
+        vectorize = not isinstance(geometry, Geometry)
+    else:
+        transform_wrapper = transform if interleaved else transform_planar
+        vectorize = False
+        if not np.isscalar(include_z):
+            if np.all(include_z == include_z[0]):
+                include_z = include_z[0]
+            else:
+                vectorize = True
+    if vectorize:
+        transform_wrapper = np.frompyfunc(transform_wrapper, 3, 1)
+
+    return transform_wrapper(
+        geometry,
+        transformation,
+        include_z,
+    )
+
+
+def transform_planar(geometry, transformation, include_z: bool = False):
```

Comment:
All these helper functions should be considered private, so prefixed with an underscore.

Reply:
Note https://github.com/shapely/shapely/pull/1676#discussion_r1147679057